modules/gpu_instance/main.tf

```diff
@@ -0,0 +1,126 @@
+/**
+ * Copyright 2019 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+locals {
+  datalab_disk_name = "${var.name}-pd"
+}
+
+/******************************************
+  Firewall rule to allow tunnel-through-iap
+ *****************************************/
+module "iap_firewall" {
+  source               = "../iap_firewall"
+  project_id           = var.project_id
+  network_name         = var.network_name
+  firewall_description = "Allow SSH and web UI IAP tunnel to Datalab instance"
+  target_tags          = ["datalab"]
+  ports                = ["22", "8080"]
+}
+
+/***********************************************
+  Render start up and cloud config scripts from templates
+ ***********************************************/
+module "template_files" {
```

Comment:
If we want to cut down on duplicated logic and variables between the submodules then the submodules could altered to accept an variable of type: 
```hcl
object({
  cloud_config = string
  startup_script = string
})
```

The calling configuration could then pass an instance of the `template_files` submodule or a different object which provides both values.

Not required right now; just some thing to consider.